Historic ask to update

aylapear opened the pull request:
- update to deal with comments in shinyrems issues https://github.com/bcgov/shinyrems/issues/126 and https://github.com/bcgov/shinyrems/issues/118
- add option that the function asks if you want to download the newer version similar to 2yr/4yr option
- kept syntax of code similar to `get_ems_data()`

aylapear committed:
updating so download historic asks if you want to update it before it takes the time to download it

aylapear committed:
styled code

aylapear committed:
adding most up to date version note back in

aylapear committed:
doc update

aylapear committed:
remove spaces

ateucher commented:
Thanks! I should be able to have a look this week

ateucher commented on R/ems_duckdb.R:
```diff
@@ -37,43 +42,86 @@ download_historic_data <- function(force = FALSE, ask = TRUE, dont_update = FALS
+  update <- FALSE
+  if (force || !db_exists) {
+    update <- TRUE # nocov
+  } else if (db_exists) {
```
I think this logic has already been covered (just in reverse) on line 42. Can they be consolidated?

aylapear replied:
I took a step back and found a way to refactor this section so it reads more clearly and is consolidated.

ateucher commented on R/ems_duckdb.R:
```diff
@@ -37,43 +42,86 @@ download_historic_data <- function(force = FALSE, ask = TRUE, dont_update = FALS
+      )
+      if (tolower(ans) == "y") update <- TRUE
+      # nocov end
+    }
```
This block doesn't check for the value of `ask` so will run regardless of its value, and will give the user a prompt, which I don't think is the intention in the case of `ask = FALSE`?

aylapear replied:
I agree with your comments here and I miss interpreted the `ask` parameter as the definition says "should the function ask for your permission to cache data on your computer?" which I took to being only about the question where it asks to save a copy of the data on your local machine. I have updated the function so the question in relation to the version is wrapped up in the ask parameter.

ateucher commented on R/ems_duckdb.R:
```diff
@@ -37,43 +42,86 @@ download_historic_data <- function(force = FALSE, ask = TRUE, dont_update = FALS
+      )
+      invisible(db_path)
+      # nocov end
+    }
```
This code is fully duplicated from the block above, unnecessarily I think.

aylapear replied:
Yes this was duplicated, I have refactored and simplified it so doesn't duplicate the code anymore

ateucher reviewed (changes requested):
Hi, thanks for this. I think I'm a little unclear what the intent is with these changes...

For point # 1 (addressing https://github.com/bcgov/shinyrems/issues/126), does using the combination of `ask = FALSE, force = FALSE`, possibly in combination with `dont_update` already accomplish that?

I've left some specific comments, but I think if you could tell me what the desired functionality is that would be helpful!

Thanks, and let me know if there's anything I can do to assist.

aylapear commented:
Sorry Andy got tied up in a couple things the last couple weeks but will review your comments and get back to you shortly. 

The intent of the changes is to make a better user experience in shiny rems. It was expressed that the historic data is now updated daily when it used to be monthly. When a user is using the app it prompts them to download the data each day now and so an option to not re-download the new data was desired similar to how the 2yr or 4yr datasets functions. If you had just downloaded it the day before you may not want to take the time to re-download the entire historic dataset. It seemed like the current options for the historic data set didn't meet those needs exactly as the other data sets do.

aylapear committed:
git merge

Merge branch 'master' into historic_ask_to_update
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

aylapear committed:
addressing the pull request questions

aylapear committed:
refactored historic data pull

aylapear commented:
Hey Andy,

Thank you for the comments as it made me take a step back and look at a better way of structuring the code. The code is organized a bit differently now so let me know if you have any further comments or suggestions. I was trying to be quick and match more of the structure of the 2yr/4yr code but it made things more confusing and bloated then just trying to structure it for the historic setup. 

This change greatly improves functionality for the shinyrems app as it allows the user to know how current their data is and whether it is worth updating it vs before you only had the option to update it if it was out of date. 
```
> shinyrems::run_ems_app("historic")
checking for historic data...
Your version of the historic ems data is dated 2022-02-16 05:57:00 and there is a newer version available. Would you like to download it? (y/n)n
Loading required package: shiny

Listening on http://127.0.0.1:5762
```

ateucher commented:
Thanks very much @aylapear for the rethinking and refactoring! It is nice and clean now - I'm just waiting for the CI to finish and then will be happy to merge!

ateucher commented on R/ems_duckdb.R:
```diff
@@ -1,79 +1,99 @@
+              ", however you have asked not to update it by setting 'dont_update' to TRUE.")
       return(invisible(db_path))
     }
+
```
```suggestion
# nocov start
```

ateucher commented on R/ems_duckdb.R:
```diff
@@ -1,79 +1,99 @@
+        )
+      )
+      if (tolower(ans) == "n") return(invisible(db_path))
+    }
```
```suggestion
    }
    # nocov end
```

ateucher commented on R/ems_duckdb.R:
```diff
@@ -1,79 +1,99 @@
+        ". Is that okay?"
+      )
+    )
   }
```
```suggestion
  }
  # nocov end
```

ateucher reviewed (changes requested):
Looks great @aylapear - just a couple of places to insert some codecov comments (since we can't test the interactive bits)

aylapear committed:
Update R/ems_duckdb.R

Co-authored-by: Andy Teucher <andy.teucher@gmail.com>

aylapear committed:
Update R/ems_duckdb.R

Co-authored-by: Andy Teucher <andy.teucher@gmail.com>

aylapear committed:
Update R/ems_duckdb.R

Co-authored-by: Andy Teucher <andy.teucher@gmail.com>

aylapear commented:
@ateucher I added those no cov lines which I was hoping will fix the issue with the failing action due to the code coverage

ateucher commented:
All green! Thanks so much @aylapear for the nice work and your patience with my slow reviews 😄
